### DISCUSSION TITLE
How to include microcode in booster image?

### DISCUSSION THREAD

**Discussion by arun54321** [anatol/booster#159] [Q&A]
Microcode is included inside dracut images like this:

```
❯ _ lsinitrd /boot/initramfs-5.18.9-200.fc36.x86_64.img | ag "\.bin"
-rw-r--r--   1 root     root       106496 Mar 30 14:23 kernel/x86/microcode/GenuineIntel.bin
```

How can I include microcode when generating initrd image using booster?

**Comment by anatol** [comment]
The recommended way (at least at Arch) is to pack microcode in a separate image and load it before the initramfs. `initrd	/boot/cpu_manufacturer-ucode.img /boot/initramfs-linux.img`. See https://wiki.archlinux.org/title/Microcode for more examples.

If you still want to put the microcode files into the image with `extra_files: /path/to/microcode` but booster itself does not do anything to load it.

**Reply by arun54321** [reply]
I use Fedora. So I cannot do the arch way.
Fedora doesn't have microcode.img file/package.

https://packages.fedoraproject.org/pkgs/microcode_ctl/microcode_ctl/fedora-36.html#files

**Comment by anatol** [comment]
A separate microcode image is a nicer cleaner/solution and Fedora/RHEL should definitely look at it.

Meantime we can look at a stopgap solution for it. Booster could detect the microcode file with code similar to following

```
get_ucode_file() {
    local family
    local model
    local stepping
    family=$(grep -E "cpu family" /proc/cpuinfo | head -1 | sed "s/.*:\ //")
    model=$(grep -E "model" /proc/cpuinfo | grep -v name | head -1 | sed "s/.*:\ //")
    stepping=$(grep -E "stepping" /proc/cpuinfo | head -1 | sed "s/.*:\ //")

    if [[ "$(get_cpu_vendor)" == "AMD" ]]; then
        if [[ $family -ge 21 ]]; then
            printf "microcode_amd_fam%xh.bin" "$family"
        else
            printf "microcode_amd.bin"
        fi
    fi
    if [[ "$(get_cpu_vendor)" == "Intel" ]]; then
        # The /proc/cpuinfo are in decimal.
        printf "%02x-%02x-%02x" "${family}" "${model}" "${stepping}"
    fi
}
```

and then put the file under `/kernel/x86/microcode/{AuthenticAMD.bin,GenuineIntel.bin}`

**Comment by anatol** [comment]
I just pushed a possible solution for this request to `wip` branch. Here is the change itself https://github.com/anatol/booster/commit/bad362484fb97c96b14361fe36169a542c023b1c

I do not have a Fedora machine and unable to test the solution properly. @arun54321 would you be able to help with testing?

**Reply by arun54321** [reply]
I'm facing issues on installing cpuidv2 dependency. I will test it after I manage to compile booster-generator.

**Comment by anatol** [comment]
I looked at my solution I like it that much. I still think that a separate ucode image is a better solution even at Fedora.

So I went ahead and made a tool to generate ucode early boot images. @arun54321 please build https://github.com/anatol/ucode-image-gen, generate the ucode image as `./ucode-image-gen`. It will produce `ucode.img` that you put into `/boot`, then modify your bootloader to include this image before the booster one.

**Reply by arun54321** [reply]
Not just fedora. Ubuntu and other popular distros use dracut, include microcode in initramfs.

I think including microcode inside initramfs is a better solution because some people use efistub loader and using a separate microcode image adds extra complexity.

**Comment by anatol** [comment]
> because some people use efistub loader

According to the Arch wiki it should be as easy as adding an extra `initrd=` option

https://wiki.archlinux.org/title/Microcode#EFISTUB
